### PR TITLE
Better Vue File Support

### PR DESCRIPTION
**Pull request by TombolaShepless**
This PR adds the new plugin `VueComponentPlugin`:

#### Features that the old `VuePlugin` does not contain:
- Adds in transpilation/pre-proessing for all blocks (`<template>`, `<script>`, `<style>`)
- Adds scoped styling functionality
- Adds ability to load external files using `src` attribute
- Has a default plugin chain for each block, but also allows per file overriding using `lang` attribute, or, global override using plugin chains
- HMR integration

I've added a deprecation warning in the old `VuePlugin` so users know to migrate over as I assume we will be removing the old plugin in due course.

Massive thanks to @nchanged and @devmondo for their help in building this plugin!

UPDATE: This will close the following: #125 #542 #821

### CODE REVIEW

**Comment by TombolaShepless** [comment]
Sure, I'll add it back in with a big console deprecation message :) gimme 5 mins.

File: package.json [hunk 1]
```diff
@@ -99,11 +99,13 @@
     "mustache": "^2.3.0",
     "opencollective": "^1.0.3",
     "postcss": "^6.0.1",
+    "postcss-selector-parser": "^2.2.3",
     "pretty-time": "^0.2.0",
     "prettysize": "0.0.3",
     "realm-utils": "^1.0.8",
     "request": "^2.79.0",
     "shorthash": "0.0.2",
+    "vue-hot-reload-api": "^2.1.0",
```
Comment (nchanged) [inline package.json]: We can't put it here, let's update the doc what's required to be installed.

File: src/index.ts [hunk 2]
```diff
@@ -12,7 +11,7 @@ export { IQuantumExtensionParams } from "./quantum/plugin/QuantumOptions";
 export { ComputedStatementRule } from "./quantum/plugin/ComputerStatementRule";
 export { QuantumPlugin } from "./quantum/plugin/QuantumPlugin";
 export { ReplacePlugin } from "./plugins/ReplacePlugin";
-export { VuePlugin } from "./plugins/VuePlugin";
```
Comment (nchanged) [inline src/index.ts]: Let's keep the old plugin! We can remove it after a few releases!

File: src/plugins/vue/VuePlugin.ts [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+import { CSSPluginClass } from "../stylesheet/CSSplugin";
+import { Concat, hashString } from "../../Utils";
+import { VueBlockFile } from './VueBlockFile';
+import { VueTemplateFile } from './VueTemplateFile';
+import { VueStyleFile } from './VueStyleFile';
+import { VueScriptFile } from './VueScriptFile';
+import * as path from "path";
+import * as fs from "fs";
+import { each } from "realm-utils";
+const vueCompiler = require("vue-template-compiler");
```
Comment (nchanged) [inline src/plugins/vue/VuePlugin.ts]: Should be hidden with the plugin. Can't be in dependencies.

Comment (TombolaShepless) [reply src/plugins/vue/VuePlugin.ts]: As in:

```js
transform(file) {
 const vueCompiler = require('vue-template-compiler')
}
```

Guess I'd need to update docs for dependency listing too

Comment (nchanged) [reply src/plugins/vue/VuePlugin.ts]: That's right! The installation must include all dependencies

File: src/plugins/vue/VuePlugin.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    cacheItem[path] = {
+      contents,
+      sourceMap
+    };
+  }
+
+  public bundleEnd(context: WorkFlowContext) {
+    if (context.useCache) {
+      context.source.addContent(`
+        var process = FuseBox.import('process');
```
Comment (nchanged) [inline src/plugins/vue/VuePlugin.ts]: Let's make sure process is included for sure. Put it into file dependencies.
And I would suggest wrapping it
```
if( process.env.NODE_ENV !== "production) {
 // that code
}
```

Comment (TombolaShepless) [reply src/plugins/vue/VuePlugin.ts]: Updated.

**Comment by eddow** [comment]
Woaw; I might have many projects to migrate but I will do it with pleasure, I'm looking forward to have time for this!
Note: it might mean "bug reports" :-p

**Pull request merged**